rpcs3/Emu/NP/rpcn_config.cpp

```diff
@@ -94,22 +128,80 @@ std::string cfg_rpcn::get_token() const
 	return token.to_string();
 }
 
-void cfg_rpcn::set_host(const std::string& host)
+void cfg_rpcn::set_host(std::string_view host)
 {
 	this->host.from_string(host);
 }
 
-void cfg_rpcn::set_npid(const std::string& npid)
+void cfg_rpcn::set_npid(std::string_view npid)
 {
 	this->npid.from_string(npid);
 }
 
-void cfg_rpcn::set_password(const std::string& password)
+void cfg_rpcn::set_password(std::string_view password)
 {
 	this->password.from_string(password);
 }
 
-void cfg_rpcn::set_token(const std::string& token)
+void cfg_rpcn::set_token(std::string_view token)
 {
 	this->token.from_string(token);
 }
+
+void cfg_rpcn::set_hosts(const std::vector<std::pair<std::string, std::string>>& vec_hosts)
+{
+	std::string final_string;
+	for (const auto& [cur_desc, cur_host] : vec_hosts)
+	{
+		fmt::append(final_string, "%s|%s|||", cur_desc, cur_host);
+	}
+
+	if (final_string.empty())
+	{
+		hosts.from_default();
+		return;
+	}
+
+	final_string.resize(final_string.size() - 3);
+	hosts.from_string(final_string);
+}
+
+bool cfg_rpcn::add_host(std::string_view new_description, std::string_view new_host)
+{
+	auto cur_hosts = get_hosts();
+
+	for (const auto& [cur_desc, cur_host] : cur_hosts)
+	{
+		if (cur_desc == new_description && cur_host == new_host)
+			return false;
+	}
+
+	cur_hosts.push_back(std::make_pair(std::string(new_description), std::string(new_host)));
+	set_hosts(cur_hosts);
+
+	return true;
+}
+
+bool cfg_rpcn::del_host(std::string_view del_description, std::string_view del_host)
+{
+	// Do not delete default servers
+	if ((del_description == "Official RPCN Server" && del_host == "np.rpcs3.net") ||
+		(del_description == "RPCN Test Server" && del_host == "test-np.rpcs3.net"))
```

Comment:
I take it these strings are all hardcoded?
wouldn't want the UI to have some translated strings.

You could use the translation callback in both get_hosts and del_host though